### PR TITLE
Adding: QnA

### PR DESCRIPTION
**Pull request by ariG23498**
As soon as the later sections of code are added I will be adding the 2 leftover QnA sections

### CODE REVIEW

File: examples/vision/probing_vits.py [hunk 1]
```diff
@@ -316,9 +316,17 @@ def compute_mean_attention_dist(patch_size, attention_weights, model_type):
 plt.show()
 
 """
-Here we can notice that different attention heads yield different attention distances
-suggesting they use both local and global information from an image. But as we go deeper
-in the transformer blocks the heads tend to focus more on global aggregate information.
+### Inspecting the plots
+
+Q. How does self-attention span across the input space? Do they attend
+input regions locally or globally?
```
Comment (sayakpaul) [inline examples/vision/probing_vits.py:322]: I don't think there's a need to append "Q." here.

File: examples/vision/probing_vits.py [hunk 2]
```diff
@@ -429,6 +437,11 @@ def attention_rollout_map(image, attention_score_dict, model_type):
 fig.show()
 
 """
+### Inspecting the plots
+
+Q. How can we quanitfy the information flow that propagates through the
+attention layers?
```
Comment (sayakpaul) [inline examples/vision/probing_vits.py:443]: Same.

File: examples/vision/probing_vits.py [hunk 3]
```diff
@@ -527,6 +540,21 @@ def attention_heatmap(attention_score_dict, image, model_type="dino"):
             axes[i, j].axis("off")
             img_count += 1
 
+"""
+### Inspecting the plots
+
+Q. How can we qualitatively evaluate the attention weights?
```
Comment (sayakpaul) [inline examples/vision/probing_vits.py:546]: Same. 

Preferably, we could put the questions in bold.

**Review by sayakpaul — CHANGES_REQUESTED** [review]
Minor changes.

File: examples/vision/probing_vits.py [hunk 4]
```diff
@@ -527,6 +540,104 @@ def attention_heatmap(attention_score_dict, image, model_type="dino"):
...
+"""
+### Inspecting the plots
+
+*What do the projection filters learn?*
+
+When visualized, the kerenls of a ConvNet speak for the pattern that
+they were looking for in a dataset. Sometimes circles, sometimes lines
+and when combined together (later stage of a ConvNet) the filters looked
+like complex shapes. We have found a stark similarity between such
+ConvNet kernels and the projection filter of a ViT.
```
Comment (ariG23498) [inline examples/vision/probing_vits.py:605]: @sayakpaul could you please review this part in particular?

I have added my thoughts here, would love to know yours as well.

Comment (sayakpaul) [reply examples/vision/probing_vits.py:605]: * `kerenls` -> `kernels`
* ConvNet -> convolutional neural network

Maybe we could add a reference link for the readers to let them explore more about what we're talking about CNN kernels. 

That said, I totally dig the philosophy you tried conveying here. That is spot on!

File: examples/vision/probing_vits.py [hunk 5]
```diff
@@ -527,6 +540,104 @@ def attention_heatmap(attention_score_dict, image, model_type="dino"):
...
+similarity = position_embeddings @ position_embeddings.T
+plt.imshow(similarity, cmap="inferno")
+plt.show()
+
+"""
+### Inspecting the plots
+
+**What do the positional embeddings tell us?**
+
+The plot has a distincive diagonal patter. The main diagonal is the brigtest
```
Comment (sayakpaul) [inline examples/vision/probing_vits.py:634]: Typo.

File: examples/vision/probing_vits.py [hunk 6]
```diff
@@ -527,6 +540,104 @@ def attention_heatmap(attention_score_dict, image, model_type="dino"):
...
+
+Transformers are permutation invariant. This means that they have
+no clue about the position of the input tokens. To overcome this
+shortcoming we inject positional information to the input tokens.
+
+The positional information can be in the form of leaned positional
+embeddings or hand-crafted constant embeddings. In our case all the
+three variants of ViTs have learned positional embeddings.
+
+In this section, we will try visualizing the similarities between the
```
Comment (sayakpaul) [inline examples/vision/probing_vits.py:619]: ..., we will visualize ...

File: examples/vision/probing_vits.py [hunk 7]
```diff
@@ -527,6 +540,104 @@ def attention_heatmap(attention_score_dict, image, model_type="dino"):
...
+"""
+### Inspecting the plots
+
+**What do the positional embeddings tell us?**
+
+The plot has a distincive diagonal patter. The main diagonal is the brigtest
+signifying that a position is the most similar to itself. An interesting
+pattern to look out for is the repeating diagonals. The repeating pattern
+portrays a sinusoidal function which is close in essence to what was proposed by
+[Vaswani et. al.](https://arxiv.org/abs/1706.03762) as a hand feature.
```
Comment (sayakpaul) [inline examples/vision/probing_vits.py:638]: ... as a hand-crafted feature

**Review by sayakpaul — CHANGES_REQUESTED** [review]
Looks good to me aside from a few minor changes.

File: examples/vision/probing_vits.py [hunk 8]
```diff
@@ -527,6 +540,105 @@ def attention_heatmap(attention_score_dict, image, model_type="dino"):
...
+no clue about the position of the input tokens. To overcome this
+shortcoming we inject positional information to the input tokens.
+
+The positional information can be in the form of leaned positional
+embeddings or hand-crafted constant embeddings. In our case all the
+three variants of ViTs have learned positional embeddings.
+
+In this section, we will visualize the similarities between the
+learned positional embeddings with itself. Below, we take the ViT B-16 model
+and visualize its similar learned positional embeddings.
```
Comment (sayakpaul) [inline examples/vision/probing_vits.py:622]: > Below, we take the ViT B-16 model and visualize its similar learned positional embeddings. 

to

> Below, we take the ViT B-16 model and visualize the similarity of the positional embeddings by taking their dot-product.

File: examples/vision/probing_vits.py [hunk 9]
```diff
@@ -527,6 +540,105 @@ def attention_heatmap(attention_score_dict, image, model_type="dino"):
...
+similarity = position_embeddings @ position_embeddings.T
+plt.imshow(similarity, cmap="inferno")
+plt.show()
+
+"""
+### Inspecting the plots
+
+**What do the positional embeddings tell us?**
+
+The plot has a distincive diagonal patter. The main diagonal is the brightest
```
Comment (sayakpaul) [inline examples/vision/probing_vits.py:635]: Typo.

* Distinctive
* Pattern

**Review by sayakpaul — APPROVED** [review]
Two minor changes. Once reflected, feel free to merge.

File: examples/vision/probing_vits.py (hunk 8 above)
Comment (ariG23498) [reply examples/vision/probing_vits.py:622]: Loved this change! Exactly what I wanted to convey but could not.

**Pull request merged**